include/envoy/stats/symbol_table.h

```diff
@@ -1,13 +1,157 @@
 #pragma once
 
+#include <memory>
+
+#include "envoy/common/pure.h"
+
+#include "absl/strings/string_view.h"
+
 namespace Envoy {
 namespace Stats {
 
-// Interface for referencing a stat name.
 class StatName;
+class SymbolEncoding;
+
+/**
+ * SymbolTable manages a namespace optimized for stats, which are typically
+ * composed of arrays of "."-separated tokens, with a significant overlap
+ * between the tokens. Each token is mapped to a Symbol (uint32_t) and
+ * reference-counted so that no-longer-used symbols can be reclaimed.
+ *
+ * We use a uint8_t array to encode a "."-deliminated stat-name into arrays of
+ * symbols in order to conserve space, as in practice the majority of token
+ * instances in stat names draw from a fairly small set of common names,
+ * typically less than 100. The format is somewhat similar to UTF-8, with a
+ * variable-length array of uint8_t. See the implementation for details.
+ *
+ * StatNameStorage can be used to manage memory for the byte-encoding. Not all
+ * StatNames are backed by StatNameStorage -- the storage may be inlined into
+ * another object such as HeapStatData. StaNameStorage is not fully RAII --
+ * instead the owner must call free(SymbolTable&) explicitly before
+ * StatNameStorage is destructed. This saves 8 bytes of storage per stat,
+ * relative to holding a SymbolTable& in each StatNameStorage object.
+ *
+ * A StatName is a copyable and assignable reference to this storage. It does
+ * not own the storage or keep it alive via reference counts; the owner must
+ * ensure the backing store lives as long as the StatName.
+ *
+ * The underlying Symbol / SymbolVec data structures are private to the
+ * impl. One side effect of the non-monotonically-increasing symbol counter is
+ * that if a string is encoded, the resulting stat is destroyed, and then that
+ * same string is re-encoded, it may or may not encode to the same underlying
+ * symbol.
+ */
+class SymbolTable {
+public:
+  /**
+   * Efficient byte-encoded storage of an array of tokens. The most common tokens
+   * are typically < 127, and are represented directly. tokens >= 128 spill into
```

Comment:
Sure, I just think the comment is a bit of a misdirect, since as a casual reader, I see this as stating that we have figured out some efficient encoding scheme and that this is an important aspect to pay attention to.